src/Sentry/Protocol/Envelopes/Envelope.cs

```diff
@@ -285,11 +282,17 @@ public static Envelope FromTransaction(Transaction transaction)
             header = CreateHeader(eventId);
         }
 
-        var items = new[]
+        var items = new List<EnvelopeItem>
         {
             EnvelopeItem.FromTransaction(transaction)
         };
 
+        if (transaction.TransactionProfiler is { } profiler)
+        {
+            // Profiler.CollectAsync() must not throw but it may complete with null in which case the EnvelopeItem won't serialize.
+            items.Add(EnvelopeItem.FromProfileInfo(profiler.CollectAsync(transaction)));
```

Comment:
In case of `null`, the `Payload` of the `EnvelopeItem` will serialize empty, but `EnvelopeItem` will still serialize.  It will have an item header showing `{"length": 0}`, and and then an empty item.  Take a close look at `EnvelopeItem.SerializeAsync`.

While you're there, you'll want to read the comments about payload buffering being synchronous.  I think for `AsyncJsonSerializable` payloads, it makes sense to go back to asynchronous buffering.  Otherwise the profile collection is actually a blocking call.

Also, we'll want some way to ensure that `Envelope.SerializeAsync` doesn't append extraneous newlines when there was nothing serialized for each item.

Reply:
> In case of `null`, the `Payload` of the `EnvelopeItem` will serialize empty, but `EnvelopeItem` will still serialize. It will have an item header showing `{"length": 0}`, and and then an empty item. Take a close look at `EnvelopeItem.SerializeAsync`.

I'm not sure what to do about that `ISerializable` is public so can't easily change that and matching specifically for `AsyncJsonSerializable` doesn't seem like the best idea... How about changing the `public async Task<ProfileInfo?> CollectAsync(...)` to a non-null task completion value `public async Task<ProfileInfo> CollectAsync(...)` and returning a default object in case of any of the invalid states? It should be an uncommon scenario anyway.

> While you're there, you'll want to read the comments about payload buffering being synchronous. I think for `AsyncJsonSerializable` payloads, it makes sense to go back to asynchronous buffering. Otherwise the profile collection is actually a blocking call.

It is blocking, but doing so in the background worker - that should be fine, shouldn't it? Bruno was OK with that back when we had a call but if you feel we should change that, please let me know (ideally with a proposal).

Reply:
If it's uncommon, then have it be non-nullable and throw a `SerializationException` if it can't be serialized instead of returning null.  Then try/catch/log that exception in envelope serialization to avoid adding the extra newline and continue with the rest of the envelope.

Reply:
> https://develop.sentry.dev/sdk/envelopes/  
> Length-prefixed payloads must terminate with \n or EOF. The newline is not considered part of the payload. Any other character, including whitespace, means the Envelope is malformed.

That implies the new line should be there regardless of the payload length.

Reply:
On the async payload serialization - I made an error in switching it to always buffer synchronously.  It should only be doing so for `JsonSerializable` payloads.  Others such as `StreamSerializable` (used for attachments) and this new `AsyncJsonSerializable` should still buffer asynchronously.  I'll make a PR for that separately so you don't need to deal with it here.

Reply:
Async buffering adjusted in #2297.

Reply:
> That implies the new line should be there regardless of the payload length.

Correct.  I'm saying that if the profile info cannot be serialized because it doesn't exist, then don't serialize an envelope item for it at all, which would include the newline.

Reply:
I've made the change to make FinishAsync() throw & also handle it during Envelope serialization so it won't have an item in that case